### ISSUE TITLE
Option to issue GET in parallel

### ISSUE THREAD

**Issue by matsken** [sagiegurari/multiple-redis#5]
First of all, thanks for creating this package - this is a perfect fit for our use case where we have two endpoints, each configured with Sentinel cluster but sometimes one of them goes down for maintenance. With this package, we can configure and forget about switching endpoints during maintenance.

**Question/problem**
If I understand correctly, GET commands are executed in series. So if there are two instances and the first instance is down, the GET command will not get sent to the 2nd one until the first one times out (after childCommandTimeout value). 

Then we have a situation where, when the first instance is down, all GET commands will take an additional X time to complete, where X is `childCommandTimeout` value.  This was quite apparent when used with redis session store, where all http requests were taking 10+ seconds.

**Describe the solution you'd like**
I would like to have an option where, if set to true, run the GET (or any) command in parallel if possible. In the case of a GET command, it will simply issue the command to all clients and respond with the first non-error result. Even if the results were different among the clients, I would be ok with that.
I imagine that it would not be too hard to implement a global option to simply issue all commands in parallel (override the `runInSequence` flag if set). What do you think?

**Comment by sagiegurari** [comment]
sounds like a good idea. will try to set it up as an option during the create.

**Comment by sagiegurari** [comment]
published new version with option: forceParallel so you force parallel for all commands.
this would increase the load on your redis servers i guess.
try it out and tell me if it works good for you

**Comment by matsken** [comment]
@sagiegurari Thanks! will try the new version.

**Comment by matsken** [comment]
@sagiegurari I've found that using `async.parallel` still waits for all async calls to finish (either with a value or an error) which is not desirable. Could we use `async.race`? With `childCommandTimeout` set to a sufficiently large value where it is greater than the non-error commands, I think we are ok.

**Comment by sagiegurari** [comment]
you are right. i'll check it out.

**Comment by sagiegurari** [comment]
updated version. hopefully this one should resolve the issue.

**Comment by sagiegurari** [comment]
@matsken did you verify?

**Comment by matsken** [comment]
@sagiegurari Sorry about the delay in response. I did try the latest version out but it did not seem to work as intended. It still seems to wait for all clients to respond with a result.

**Comment by sagiegurari** [comment]
are you sure you are passing the new flag?
i have a test (https://github.com/sagiegurari/multiple-redis/blob/master/test/spec/multiple-redis-spec.js#L850) specifically to test this and it would fail if it had to wait for all clients to finish.

**Comment by matsken** [comment]
Yes, I'm sure I have this set correctly:
```javascript
const multiClient = MultipleRedis.createClient(clients, {childCommandTimeout: 2000, forceParallel: true});
```

Did a little testing myself, and found that this issue only happens when a `get` call is invoked on a key that does not exist (= should return null). When the value is explicitly set to null for the key, it works.

Sample code: 
```javascript
const MultipleRedis = require("multiple-redis");
const redis = require("redis");

const clients = [{host: "localhost"}, {host: "localhost", port: 6380}].map(opt => redis.createClient(opt));

const client = MultipleRedis.createClient(clients, {childCommandTimeout: 2000, forceParallel: true});

client.set("key1", null);
const start = Date.now();

setInterval(() => {
	client.get("key1", (err, res) => {
		const time = Date.now() - start;
		console.log(`key1: ${time}ms: ${res}`);
	});
	client.get("key2", (err, res) => {
		const time = Date.now() - start;
		console.log(`key2: ${time}ms: ${res}`);
	});
}, 1000);
```

Result:
```
key1: 1002ms: null
key1: 2003ms: null
key2: 3004ms: null
key1: 3004ms: null
key2: 4005ms: null
key1: 4006ms: null
key2: 5006ms: null
key1: 5006ms: null
```

See that the key2 value only starts appearing at 3000ms, which is `childCommandTimeout` + `setInterval` time.

**Comment by sagiegurari** [comment]
so issue happens when all redis servers basically fail to respond. got it.

**Comment by matsken** [comment]
I did notice that changing `async.parallel` to `async.race` achieves what I wanted, but you probably had your reasons to not use `race`.

**Comment by sagiegurari** [comment]
race will stop on first error so if second redis would return results we would not get them

**Comment by matsken** [comment]
@sagiegurari So what do you suggest we do? I would be OK with `race` in that error will arrive before a result, because I can set the `childCommandTimeout` to a sufficiently large value to avoid this condition. Or am I misunderstanding something...?

**Comment by sagiegurari** [comment]
@matsken race does not solve your issue.
you have 2 redis servers, 1 always throws an error fast, the other has a result but takes a bit of time to return. with race, you will only get errors although one of the redis servers got the data you need.

**Comment by matsken** [comment]
@sagiegurari Not sure if I'm seeing what you're seeing though - when I configure 2 clients and have 1 instance offline, using `race` works as I expect, as in, the values are always set/get on the live instance - it works as if only one instance is configured.  Maybe the timeout issue is happening in situations where the client does not respond with an error. Under what condition would a client wait until `childCommandTimeout` expires?

**Comment by sagiegurari** [comment]
* from the docs - http://caolan.github.io/async/v3/docs.html#race
> Runs the tasks array of functions in parallel, without waiting until the previous function has completed. Once any of the tasks complete or pass an error to its callback, the main callback is immediately called. It's equivalent to Promise.race().

Therefore, if one redis has the value to give, but another redis returned an error first, you will get an error. that is bad.

* childCommandTimeout - will be invoked in a get function in case all redis servers didn't return an answer or i got null/empty value. i think this is your issue, that your redis servers are returning nothing and it causes you to wait. that is what i think i need to modify

**Comment by matsken** [comment]
Right, that is what's happening. I think if an instance is down, instead of an error being returned immediately, the redis client waits for a timeout.

**Comment by sagiegurari** [comment]
so you will be stuck until timeout only if all instances are down...

**Comment by matsken** [comment]
@sagiegurari Actually, that's not what I'm seeing - if at least one of the clients is waiting for timeout, the entire callback seems to wait for timeout, even if one of the clients immediately returns with a value.

**Comment by sagiegurari** [comment]
ok, will try to reproduce. seems different in my tests so want to be sure

**Comment by sagiegurari** [comment]
i'm just not sure why you are seeing that, you will have to start debugging.
the below is a test case i wrote that tests 5 'fake' redis servers. only the last provides an answer.
the rest either return nothing, timeout or return an error.
the test runs super fast without waiting for a timeout (see that i setup mocha to timeout after 100 millies) and with the actual answer.
so try to debug and see what is going on...

```js
it('last has data force parallel, others timeout/error/null', function (done) {
    this.timeout(100);

    var count = 0;
    var createClient = function () {
        return {
            on: noop,
            send_command: function (name, args, callback) {
                count++;

                assert.equal(name, 'get');
                assert.deepEqual(args, ['my key']);
                assert.isFunction(callback);

                var currentCount = count;

                setTimeout(function () {
                    switch (currentCount) {
                    case 1:
                        setTimeout(callback, 120000);
                        break;
                    case 2:
                        callback(null, null);
                        break;
                    case 3:
                        callback(new Error('test'));
                        break;
                    case 4:
                        callback();
                        break;
                    default:
                        callback(null, 'my value');
                    }
                }, 1);
            }
        };
    };
    var client1 = createClient();
    var client2 = createClient();
    var client3 = createClient();
    var client = MultipleRedis.createClient([
        createClient(),
        createClient(),
        createClient(),
        createClient(),
        createClient()
    ], {
        forceParallel: true
    });

    assert.isTrue(client.forceParallel);

    client.get('my key', function (error, response) {
        assert.isNull(error);
        assert.equal(response, 'my value');
        assert.equal(count, 5);

        done();
    });
});
```

**Comment by matsken** [comment]
@sagiegurari OK, I did a bit more verbose debugging and found out that the culprit was the `expire` command issued by `connect-redis` which is invoked whenever session object touched. Since this is not a get command, this is processed in series and waits for timeout expiration. So we would need to allow parallel calls for not only `get` but also other commands such as `expire`.  Could we allow all commands to be issued in parallel if the `forceParallel` option is set?

**Comment by sagiegurari** [comment]
set operations have always ran in parallel. never in series.
but i do wait for all servers to finish, as you do not want to do a get before you finish the set.

from the code:
```js
var runInParallel = !getCommand && (command.indexOf('setnx') === -1);
```
all non get commands (except setnx) run in parallel and you don't have any option to run in series.

**Comment by matsken** [comment]
Could we loosen this restriction if `forceParallel` is set to true, or would it become unusable if we allow sets to return as soon as one client finishes?

**Comment by sagiegurari** [comment]
i wouldn't recommend that.
think of following flow: you set the session with lets say auth=true. then after only first redis was done it ends and another request comes and you do get but second redis answers first and you get auth=false and throw user out. that's a problem (the reverse with the auth would be huge security issue).
so... set should return until we are done with all servers.

**Comment by matsken** [comment]
OK, then what about allowing parallel for `expire` only? This will make it usable with connect-redis, since that is the command that causes the wait (=for timeout).

**Comment by sagiegurari** [comment]
@matsken seems i completely forgot about this issue, sorry.
where are we with this? do you still need it? do you want to push a PR or do you need me to publish this change
